### ISSUE TITLE
[Bug]: aws_efs_file_system_policy always applies changes

### ISSUE THREAD

**Issue by sinkr** [hashicorp/terraform-provider-aws#32077]
### Terraform Core Version

1.5.0

### AWS Provider Version

5.4.0

### Affected Resource(s)

* aws_efs_file_system_policy continually applies changes

### Expected Behavior

Application should occur once and simultaneous `terraform apply` runs should run clean with no changes necessary for for `aws_efs_file_system_policy`

### Actual Behavior

Every single run, 

`  statement {
    sid    = "efs-statement-deny"
    effect = "Deny"
`

Is applied

### Relevant Error/Panic Output Snippet

```shell
# module.eks.aws_efs_file_system_policy.policy will be updated in-place
  ~ resource "aws_efs_file_system_policy" "policy" {
        id                                 = "fs-<redacted>"
      ~ policy                             = jsonencode(
          ~ {
              ~ Statement = [
                    # (1 unchanged element hidden)
                    {
                        Action    = "*"
                        Condition = {
                            Bool = {
                                "aws:SecureTransport" = "false"
                            }
                        }
                        Effect    = "Deny"
                        Principal = {
                            AWS = "*"
                        }
                        Resource  = "arn:aws:elasticfilesystem:us-west-2:<redacted>:file-system/fs-<redacted>"
                        Sid       = "efs-statement-deny"
                    },
                  ~ {
                      - Resource  = "arn:aws:elasticfilesystem:us-west-2:<redacted>:file-system/fs-<redacted>"
                        # (5 unchanged attributes hidden)
                    },
                  ~ {
                      - Resource  = "arn:aws:elasticfilesystem:us-west-2:<redacted>:file-system/fs-<redacted>"
                        # (4 unchanged attributes hidden)
                    },
                ]
                # (1 unchanged attribute hidden)
            }
        )
        # (2 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

### Terraform Configuration Files

resource "aws_efs_file_system_policy" "policy" {
  file_system_id                     = data.terraform_remote_state.vpc.outputs.eks_efs_id
  policy                             = data.aws_iam_policy_document.policy.json
  bypass_policy_lockout_safety_check = true
}

data "aws_iam_policy_document" "policy" {

  statement {
    sid    = "efs-statement-allow"
    effect = "Allow"

    principals {
      type        = "AWS"
      identifiers = ["*"]
    }

    actions = [
      "elasticfilesystem:ClientRootAccess",
      "elasticfilesystem:ClientWrite"
    ]

    resources = ["arn:aws:elasticfilesystem:${var.aws_region}:${data.aws_caller_identity.current_account.id}:file-system/${data.terraform_remote_state.vpc.outputs.eks_efs_id}"]

    condition {
      test     = "Bool"
      variable = "elasticfilesystem:AccessedViaMountTarget"
      values   = ["true"]
    }
  }

  statement {
    sid    = "efs-statement-deny"
    effect = "Deny"

    principals {
      type        = "AWS"
      identifiers = ["*"]
    }

    actions = ["*"]

    resources = ["arn:aws:elasticfilesystem:${var.aws_region}:${data.aws_caller_identity.current_account.id}:file-system/${data.terraform_remote_state.vpc.outputs.eks_efs_id}"]

    condition {
      test     = "Bool"
      variable = "aws:SecureTransport"
      values   = ["false"]
    }
  }

  statement {
    sid    = "efs-statement-allow-irsa"
    effect = "Allow"

    principals {
      type = "Federated"
      identifiers = [format(
        "arn:aws:iam::${data.aws_caller_identity.current_account.id}:%s",
        replace(
          "${module.eks.cluster_oidc_issuer_url}",
          "https://",
          "oidc-provider/"
        )
      )]
    }

    actions = ["elasticfilesystem:*"]

    condition {
      test     = "StringLike"
      variable = "${trimprefix(module.eks.cluster_oidc_issuer_url, "https://")}:sub"
      values   = ["system:serviceaccount:kube-system:efs-csi-*"]
    }
  }

  statement {
    sid    = "efs-statement-allow-iam"
    effect = "Allow"

    principals {
      type        = "AWS"
      identifiers = ["arn:aws:iam::${data.aws_caller_identity.current_account.account_id}:role/efs-csi-provisioner"]
    }

    actions = ["elasticfilesystem:*"]
  }
}

### Steps to Reproduce

1. Apply the EFS FS policy
2. Attempt to reapply the FS policy

### Would you like to implement a fix?

None

**Comment by justinretzolk** [comment]
Hey @sinkr 👋 Thank you for taking the time to raise this! Looking out the plan output, it doesn't appear that the statement with SID of `efs-statement-deny` is being added (it doesn't have the `~` indicating a change) but rather that the `Resource` argument of two of the other statements are being removed (notice the `-` next to `Resource`), perhaps due to the statements with SIDs of `efs-statement-allow-irsa` and `efs-statement-allow-iam` not having their `resources` argument defined.

**Comment by sinkr** [comment]
Hi @justinretzolk, thank you for the response, but I'm not sure I follow.  What I end up with is the correct policy in AWS that works and functions correctly for EFS (secure) mounts to the EKS cluster only.  The redacted policy, directly from the AWS editor is below as injected by the Terraform above:
<img width="836" alt="image" src="https://github.com/hashicorp/terraform-provider-aws/assets/696374/3026ee95-2bb1-4535-9e55-00888cfe6fc9">

**Comment by trevorrea** [comment]
I'm not entirely sure what Justin is saying either but just to add that I have seen this behaviour as well where Terraforn is constantly flagging a diff even when the policy document is not changing when trying to use a aws_iam_policy_document data resource to construct an EFS file system policy so it's not just @sinkr seeing it.

In the end I got fed up with the constant diff and just used a HEREDOC to add the policy directly as json which isn't really a solution!

TL:DR - I've seen this too but don't know why it happens either.

**Comment by justinretzolk** [comment]
Hey @sinkr 👋 Thank you for that screenshot -- that's very helpful in confirming what I thought (and sorry that I clearly wasn't clear enough in what I was trying to say!). 

The `aws_iam_policy_document` data source can take multiple `statement` blocks, which are merged together to create the array of `Statement` objects in the resulting JSON. When looking at your initial log output, you mentioned that the statement with a SID of `efs-statement-deny` was being added every time, however, my read on that log is different.

While the plan log shows a change to the `Statement` array in the JSON (indicated by the `~` next to `Statement`) this is indicating that there's a change in the `Statement` array. The exact change is actually a bit further down in the JSON, where there are two lines that are being removed (indicated by the `-` next to them):

```shell
                  ~ {
                      - Resource  = "arn:aws:elasticfilesystem:us-west-2:<redacted>:file-system/fs-<redacted>"
                        # (5 unchanged attributes hidden)
                    },
                  ~ {
                      - Resource  = "arn:aws:elasticfilesystem:us-west-2:<redacted>:file-system/fs-<redacted>"
                        # (4 unchanged attributes hidden)
                    },
```

What I'd intended to get across in my initial comment was that these two `Resource` lines are what are changing in the plan; the statement with a SID of `efs-statement-deny` is unchanged, based on that plan output.

In reviewing the Terraform configuration that you supplied, there are two statements that do not have a `resource` argument in the `aws_iam_policy_document` data source with SIDs of `efs-statement-allow-irsa` and `efs-statement-allow-iam`:

```terraform
  statement {
    sid    = "efs-statement-allow-irsa"
    effect = "Allow"

    principals {
      type = "Federated"
      identifiers = [format(
        "arn:aws:iam::${data.aws_caller_identity.current_account.id}:%s",
        replace(
          "${module.eks.cluster_oidc_issuer_url}",
          "https://",
          "oidc-provider/"
        )
      )]
    }

    actions = ["elasticfilesystem:*"]

    condition {
      test     = "StringLike"
      variable = "${trimprefix(module.eks.cluster_oidc_issuer_url, "https://")}:sub"
      values   = ["system:serviceaccount:kube-system:efs-csi-*"]
    }
  }

  statement {
    sid    = "efs-statement-allow-iam"
    effect = "Allow"

    principals {
      type        = "AWS"
      identifiers = ["arn:aws:iam::${data.aws_caller_identity.current_account.account_id}:role/efs-csi-provisioner"]
    }

    actions = ["elasticfilesystem:*"]
  }
```

Looking at the screenshot that you provided from the AWS policy editor, both of these statements _do_ have a `Resource` key in the policy. This means that when Terraform reads the policy back, it will see a diff, and attempt to remove the `Resource` block, which is what we're seeing in the plan output that you initially posted.

<img width="836" alt="247215104-3026ee95-2bb1-4535-9e55-00888cfe6fc9" src="https://github.com/hashicorp/terraform-provider-aws/assets/44710313/b997bb56-7bca-4e91-885b-2a859aeec066">

**Comment by sinkr** [comment]
Hi @justinretzolk, thank you for the clarification.  I see that this is not a bug, but instead PEBCAK. 

The fix is to add the resource to the missing `Statement`, which produce a clean plan.

Thank you so much
